### PR TITLE
Improvements made to Weluse's RBAC adapter and the li3_access library.

### PR DESCRIPTION
**Pull request by rich97**
I wanted to use weluse's RBAC adapter but it was missing a few things I needed and was lacking tests.

I've given the original adapter a bit of an overhaul and made a few small changes to the li3_access library.

Changes to RBAC:
- Uses named parameters in the roles configuration.
- 100% unit tested.
- Allows you to change the message and redirect options on a per-rule basis. See the documentation for more information.
- Documentation improvements.

Changes to li3_access:
- Cleaned up the documentation and added some formatting to it.
- Ensured that all tests pass.
- Added tests for Simple adapter. I know this isn't really necessary but it was bugging me that tests didn't exist for it.

There are a couple of small issues with the new Rbac adapter but they are small and I plan to tackle them soon.

Regards,
Richard.

### CODE REVIEW

**Comment by rich97** [comment]
I've just added the ability to pass an anonymous function as a match condition. There is a bug in it which causes the role message and redirect to not be honoured but it works for the most part. I just need to think of a solution which doesn't involve me changing how the API works cause it's going to be a pain to rewrite all of those tests.

``` php
Access::config(
    array('rbac' => array(
        'adapter' => 'AuthRbac',
        'roles' => array(
            array(
                'allow' => false,
                'requesters' => '*',
                'redirect' => 'Access::login',
                'match' => array('library' => 'li3_admin', '*::*')
            ), 
            array(
                'requesters' => '*',
                'match' => array('library' => 'li3_admin', 'Access::*')
            ), 
            array(
                'requesters' => '*',
                'redirect' => 'Access::login',
                'message' => 'Already setup dude!',
                'match' => array(
                    function() {
                        // This must return true in order for this condition to match pass
                        // In this particular example I want to check if the system has already been setup, if it has then I return false.
                        return false;
                    }, 
                    'library' => 'li3_admin',
                    'Settings::setup'
                )  
            )  
        )  
    )) 
);
```

**Comment by m4rcsch** [comment]
looks quite nice. will test it soon :)

**Comment by rich97** [comment]
OK It's back to working order now. :)

**Comment by m4rcsch** [comment]
Great. Will check it out!

File: security/AccessDeniedException.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * Lithium: the most rad php framework
+ *
+ * @copyright     Copyright 2011, Union of RAD (http://union-of-rad.org)
+ * @license       http://opensource.org/licenses/bsd-license.php The BSD License
+ */
+
+namespace li3_access\security;
+
+/**
+ * A `NetworkException` may be thrown whenever an unsuccessful attempt is made to connect to a
```
Comment (nateabele) [inline security/AccessDeniedException.php:12]: This docblock should probably be fixed. :-)

**Pull request merged**